java/ql/src/experimental/Security/CWE/CWE-094/TemplateInjection.qll

```diff
@@ -0,0 +1,247 @@
+  }
+
+  /**
+   * An argument to FreeMarker template engine's `process` method call is a sink for `Configuration`.
+   */
+  private class FreeMarkerProcessSink extends Sink {
+    FreeMarkerProcessSink() {
+      exists(MethodAccess m |
+        m.getCallee() instanceof MethodFreeMarkerTemplateProcess and
+        m.getArgument(0) = this.getExpr()
+      )
+    }
+  }
+
+  /**
+   * An reader passed an argument to FreeMarker template engine's `Template`
+   *   construtor call is a sink for `Configuration`.
+   */
+  private class FreeMarkerConstructorSink extends Sink {
+    FreeMarkerConstructorSink() {
+      // Template(java.lang.String name, java.io.Reader reader)
+      // Template(java.lang.String name, java.io.Reader reader, Configuration cfg)
+      // Template(java.lang.String name, java.io.Reader reader, Configuration cfg, java.lang.String encoding)
+      // Template(java.lang.String name, java.lang.String sourceCode, Configuration cfg)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg, ParserConfiguration customParserConfiguration, java.lang.String encoding)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg, java.lang.String encoding)
+      exists(ConstructorCall cc, Expr e |
+        cc.getConstructor().getDeclaringType() instanceof TypeFreeMarkerTemplate and
+        e = cc.getAnArgument() and
+        e.getType().(RefType).hasQualifiedName("java.io", "Reader") and
+        this.asExpr() = e
+      )
+    }
+  }
+
+  /**
+   * An argument to FreeMarker template engine's `putTemplate` method call is a sink for `Configuration`.
+   */
+  private class FreeMarkerStringTemplateLoaderPutTemplateSink extends Sink {
+    FreeMarkerStringTemplateLoaderPutTemplateSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(1) and
+        ma.getMethod() instanceof MethodFreeMarkerStringTemplateLoaderPutTemplate
+      )
+    }
+  }
+
+  /**
+   * An argument to Pebble template engine's `getLiteralTemplate` method call is as a sink for `Configuration`.
+   */
+  private class PebbleGetTemplateSinkTemplateSink extends Sink {
+    PebbleGetTemplateSinkTemplateSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(_) and
+        ma.getMethod() instanceof MethodPebbleGetTemplate
+      )
+    }
+  }
+
+  /**
+   * An argument to JinJava template engine's `render` or `renderForResult` method call is as a sink for `Configuration`.
+   */
+  private class JinjavaRenderSink extends Sink {
+    JinjavaRenderSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(_) and
+        ma.getMethod() instanceof MethodJinjavaRenderForResult
+        or
+        ma.getMethod() instanceof MethodJinjavaRender
```

Comment:
`this` is not bounded on the MethodJinjavaRender arm of the `or` statement.

I suspect you might want:

```suggestion
        this.asExpr() = ma.getArgument(_) and
        (ma.getMethod() instanceof MethodJinjavaRenderForResult or ma.getMethod() instanceof MethodJinjavaRender)
```

(or whatever the autoformatter will make of that)